encoder/v1/encoder.go

```diff
@@ -0,0 +1,218 @@
+package v1
+
+import (
+	"fmt"
+	"math/big"
+	"strings"
+
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	encoders "github.com/smartcontractkit/ocr2keepers/encoder"
+	"github.com/smartcontractkit/ocr2keepers/pkg/chain"
+	"github.com/smartcontractkit/ocr2keepers/pkg/types"
+)
+
+const (
+	separator = "|"
+)
+
+var (
+	errUpkeepKeyNotParsable = fmt.Errorf("upkeep key not parsable")
+)
+
+type wrappedPerform struct {
+	CheckBlockNumber uint32   `abi:"checkBlockNumber" json:"checkBlockNumber"`
+	CheckBlockhash   [32]byte `abi:"checkBlockhash" json:"checkBlockhash"`
+	PerformData      []byte   `abi:"performData" json:"performData"`
+}
+
+type reportPacker interface {
+	Pack(args ...interface{}) ([]byte, error)
+	UnpackIntoMap(v map[string]interface{}, data []byte) error
+}
+
+type encoder struct {
+	reportKeys []string
+	packer     reportPacker
+}
+
+func NewEncoder() *encoder {
+	arguments := abi.Arguments{
```

Comment:
Right now the constructor defines the ABI arguments, but we could allow the caller to specify and pass in the arguments instead? But I'm not sure how different versions of the encoder might behave so I've kept it like this for now

Reply:
A pattern I was toying with was using composure a bit more.

```
package encode

type BigIntBlockKey struct {}

func (k BigIntBlockKey) ValidateBlockKey(BlockKey) (bool, error) {
    // ... apply validation logic
}
```

Which allows for something like this to be done elsewhere:

```
package somethingelse

type V1ContractEncoder struct {
    encode.BigIntBlockKey
    // ... other composed structs
}

func (e *V1ContractEncoder) EncodeReport(UpkeepResult, Config) ([]byte, error) {
    // ... implementation for custom contract
    // it is far more likely that each contract will have a different report encoding
    // than each contract having different BlockKey validation
}
```